File: Python/Polymorphism.py
from typing import List, Optional
from dataclasses import dataclass, field


@dataclass
class Player:
    name: str = ""
    address: str = ""


@dataclass
class Team:
    players: List[Player] = field(default_factory=list)

    def add_player(self, p: Player) -> None:
        self.players.append(p)

    def remove_player(self, p: Player) -> None:
        self.players.remove(p)


@dataclass
class SponsorCompany:
    player: Optional[Player] = None

    def hire_player(self, p: Player) -> None:
        self.player = p


# Composition

@dataclass
class File:
    file_name: str = ""
    content: str = ""
    size: int = 0


@dataclass
class Folder:
    files: List[File] = field(default_factory=list)

    def touch(self, file_name: str) -> None:
        self.files.append(File(file_name))

File: Golang/Polymorphism.go
package main

type Player struct {
	Name    string
	Address string
}

type Team struct {
	Players []*Player
}

func (t *Team) AddPlayers(p *Player) {
	t.Players = append(t.Players, p)
}

func (t *Team) RemovePlayer(p *Player) {
	for i, player := range t.Players {
		if player == p {
			t.Players = append(t.Players[:i], t.Players[i+1:]...)
			break
		}
	}
}

type SponsorCompany struct {
	Player *Player
}

func (s *SponsorCompany) HirePlayer(p *Player) {
	s.Player = p
}

// Composition

type File struct {
	FileName string
	Content  string
	Size     int
}

type Folder struct {
	Files []*File
}

func (fo *Folder) Touch(fileName string) {
	fo.Files = append(fo.Files, &File{FileName: fileName})
}

func main() {}
